File: api/src/main/java/io/github/runtimemodels/chazm/id/Identifiable.java
package io.github.runtimemodels.chazm.id;

import io.github.runtimemodels.chazm.Organization;

/**
 * The {@linkplain Identifiable} interface is used to mark elements of an {@linkplain Organization} that can be uniquely identified.
 *
 * @param <T> the type of the {@linkplain UniqueId}.
 * @author Christopher Zhong
 * @since 7.0.0
 */
@FunctionalInterface
public interface Identifiable<T> {
    /**
     * Returns the {@linkplain UniqueId}.
     *
     * @return the {@linkplain UniqueId}.
     */
    UniqueId<T> getId();
}

File: chazm-api/src/main/kotlin/runtimemodels/chazm/api/id/Identifiable.kt
package runtimemodels.chazm.api.id

import runtimemodels.chazm.api.organization.Organization

/**
 * The [Identifiable] interface is used to mark entities of an [Organization] that can be uniquely identified.
 *
 * @param <T> the type for the [UniqueId].
 * @param <U> the subtype of the [UniqueId].
 *
 * @author Christopher Zhong
 * @since 7.0.0
 */
@FunctionalInterface
interface Identifiable<T : Any, U : UniqueId<T>> {
    /**
     * The `id`.
     */
    val id: U
}
